File: tests/wasm/test_gaussian.js
/**
 * WASM binding tests for Gaussian indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import {
  loadTestData,
  assertArrayClose,
  assertClose,
  isNaN,
  assertAllNaN,
  assertNoNaN,
  EXPECTED_OUTPUTS,
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
  
  try {
    const wasmPath = path.join(__dirname, '../../pkg/my_project.js');
    const importPath = process.platform === 'win32' ? 'file:///' + wasmPath.replace(/\\/g, '/') : wasmPath;
    wasm = await import(importPath);
  } catch (error) {
    console.error('Failed to load WASM module. Run "wasm-pack build -- --features wasm --no-default-features" first');
    throw error;
  }

  testData = loadTestData();
});

test('Gaussian accuracy', async () => {
  const close = new Float64Array(testData.close);
  const expected = EXPECTED_OUTPUTS.gaussian;

  const result = await wasm.gaussian_js(close, expected.defaultParams.period, expected.defaultParams.poles);

  assert.strictEqual(result.length, close.length);

  
  const last5 = result.slice(-5);
  assertArrayClose(last5, expected.last5Values, 1e-4, 'Gaussian last 5 values mismatch');

  
  await compareWithRust('gaussian', result, 'close', expected.defaultParams, 1e-10);
});

test('Gaussian zero period', () => {
  const data = new Float64Array([10, 20, 30]);
  assert.throws(() => wasm.gaussian_js(data, 0, 4));
});

test('Gaussian period exceeds length', () => {
  const data = new Float64Array([10, 20, 30]);
  assert.throws(() => wasm.gaussian_js(data, 10, 4));
});

test('Gaussian very small dataset', () => {
  const data = new Float64Array([42]);
  assert.throws(() => wasm.gaussian_js(data, 14, 4));
});

test('Gaussian empty input', () => {
  const data = new Float64Array([]);
  assert.throws(() => wasm.gaussian_js(data, 14, 4));
});

test('Gaussian invalid poles', () => {
  const data = new Float64Array([1, 2, 3, 4, 5]);
  assert.throws(() => wasm.gaussian_js(data, 3, 0)); 
  assert.throws(() => wasm.gaussian_js(data, 3, 5)); 
});

test('Gaussian period one degeneracy', () => {
  const data = new Float64Array([1, 2, 3, 4, 5]);
  assert.throws(() => wasm.gaussian_js(data, 1, 2));
});

test('Gaussian NaN handling', () => {
  const close = new Float64Array(testData.close);
  const res = wasm.gaussian_js(close, 14, 4);
  assert.strictEqual(res.length, close.length);
  
  const skip = 4; 
  for (let i = skip; i < res.length; i++) {
    assert(Number.isFinite(res[i]), `Non-finite value at index ${i}`);
  }
});

test('Gaussian batch single parameter set', () => {
  const close = new Float64Array(testData.close);
  const single = wasm.gaussian_js(close, 14, 4);
  const batch = wasm.gaussian_batch(close, { period_range: [14, 14, 0], poles_range: [4, 4, 0] });

  assert.strictEqual(batch.values.length, single.length);
  assert.strictEqual(batch.rows, 1);
  assert.strictEqual(batch.cols, close.length);
  assertArrayClose(batch.values, single, 1e-10, 'Batch vs single mismatch');
});

test('Gaussian batch multiple combinations + metadata', () => {
  const close = new Float64Array(testData.close.slice(0, 50));
  const result = wasm.gaussian_batch(close, { period_range: [10, 20, 5], poles_range: [2, 4, 1] });

  
  assert.strictEqual(result.rows, 9);
  assert.strictEqual(result.cols, 50);
  assert.strictEqual(result.values.length, 9 * 50);
  assert.strictEqual(result.combos.length, 9);

  
  const idx = result.combos.findIndex(c => c.period === 15 && c.poles === 3);
  assert.ok(idx >= 0, 'Expected combo (15,3) missing');
  const row = result.values.slice(idx * result.cols, (idx + 1) * result.cols);
  const single = wasm.gaussian_js(close, 15, 3);
  assertArrayClose(row, single, 1e-10, 'Batch row mismatch for (period=15,poles=3)');
});

test('Gaussian zero-copy API (alloc/into/free)', () => {
  const data = new Float64Array(Array.from({ length: 64 }, (_, i) => i + 1));
  const period = 10;
  const poles = 3;

  
  const ptr = wasm.gaussian_alloc(data.length);
  assert(ptr !== 0, 'Failed to allocate memory');

  const memory = wasm.__wbindgen_memory ? wasm.__wbindgen_memory() : wasm.memory;
  if (!memory || !memory.buffer) {
    console.warn('Skipping zero-copy API test: wasm memory accessor unavailable');
    wasm.gaussian_free(ptr, data.length);
    return;
  }

  const view = new Float64Array(memory.buffer, ptr, data.length);
  view.set(data);

  try {
    
    wasm.gaussian_into(ptr, ptr, data.length, period, poles);
    const regular = wasm.gaussian_js(data, period, poles);
    for (let i = 0; i < data.length; i++) {
      if (isNaN(regular[i]) && isNaN(view[i])) continue;
      assert(Math.abs(regular[i] - view[i]) < 1e-10, `Zero-copy mismatch at ${i}`);
    }
  } finally {
    wasm.gaussian_free(ptr, data.length);
  }
});

File: tests/python/test_gaussian.py
"""
Python binding tests for Gaussian indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:

    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close
from rust_comparison import compare_with_rust


class TestGaussian:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()

    def test_gaussian_partial_params(self, test_data):
        """Test Gaussian with partial parameters - mirrors check_gaussian_partial_params"""
        close = test_data['close']


        result = ta_indicators.gaussian(close, 14, 4)
        assert len(result) == len(close)

    def test_gaussian_accuracy(self, test_data):
        """Test Gaussian matches expected values from Rust tests - mirrors check_gaussian_accuracy"""
        close = test_data['close']


        result = ta_indicators.gaussian(close, 14, 4)

        assert len(result) == len(close)


        expected_last_five = [
            59221.90637814869,
            59236.15215167245,
            59207.10087088464,
            59178.48276885589,
            59085.36983209433
        ]


        assert_close(
            result[-5:],
            expected_last_five,
            rtol=1e-4,
            msg="Gaussian last 5 values mismatch"
        )


        compare_with_rust('gaussian', result, 'close', {'period': 14, 'poles': 4})

    def test_gaussian_default_candles(self, test_data):
        """Test Gaussian with default parameters - mirrors check_gaussian_default_candles"""
        close = test_data['close']


        result = ta_indicators.gaussian(close, 14, 4)
        assert len(result) == len(close)


        compare_with_rust('gaussian', result, 'close', {'period': 14, 'poles': 4})

    def test_gaussian_zero_period(self):
        """Test Gaussian fails with zero period - mirrors check_gaussian_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])

        with pytest.raises(ValueError):
            ta_indicators.gaussian(input_data, period=0, poles=4)

    def test_gaussian_period_exceeds_length(self):
        """Test Gaussian fails when period exceeds data length - mirrors check_gaussian_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0])

        with pytest.raises(ValueError):
            ta_indicators.gaussian(data_small, period=10, poles=4)

    def test_gaussian_very_small_dataset(self):
        """Test Gaussian with very small dataset - mirrors check_gaussian_very_small_dataset"""
        data_single = np.array([42.0])

        with pytest.raises(ValueError):
            ta_indicators.gaussian(data_single, period=3, poles=4)

    def test_gaussian_empty_input(self):
        """Test Gaussian with empty input - mirrors check_gaussian_empty_input"""
        data_empty = np.array([])

        with pytest.raises(ValueError):
            ta_indicators.gaussian(data_empty, period=14, poles=4)

    def test_gaussian_invalid_poles(self):
        """Test Gaussian with invalid poles - mirrors check_gaussian_invalid_poles"""
        data = np.array([1.0, 2.0, 3.0, 4.0, 5.0])


        with pytest.raises(ValueError):
            ta_indicators.gaussian(data, period=3, poles=0)


        with pytest.raises(ValueError):
            ta_indicators.gaussian(data, period=3, poles=5)

    def test_gaussian_period_one(self):
        """Test Gaussian fails with period=1 (degenerate case) - mirrors check_gaussian_period_one_degenerate"""
        data = np.array([1.0, 2.0, 3.0, 4.0, 5.0])

        with pytest.raises(ValueError, match="Period of 1 causes degenerate"):
            ta_indicators.gaussian(data, period=1, poles=2)

    def test_gaussian_all_nan(self):
        """Test Gaussian with all NaN input - mirrors check_gaussian_all_nan"""
        data = np.array([np.nan, np.nan, np.nan, np.nan, np.nan])

        with pytest.raises(ValueError):
            ta_indicators.gaussian(data, period=3, poles=4)

    def test_gaussian_reinput(self, test_data):
        """Test Gaussian with re-input of Gaussian result - mirrors check_gaussian_reinput"""
        close = test_data['close']


        first_result = ta_indicators.gaussian(close, 14, 4)


        second_result = ta_indicators.gaussian(first_result, 10, 2)

        assert len(second_result) == len(first_result)


        for i in range(240, len(second_result)):
            assert not np.isnan(second_result[i]), f"NaN found at index {i}"

    def test_gaussian_nan_handling(self, test_data):
        """Test Gaussian handling of NaN values - mirrors check_gaussian_nan_handling"""


        close = test_data['close']

        result = ta_indicators.gaussian(close, period=14, poles=4)

        assert len(result) == len(close)

        skip = 4
        for i in range(skip, len(result)):
            assert np.isfinite(result[i]), f"Non-finite value found at index {i}"

    def test_gaussian_warmup_period(self, test_data):
        """Test that Gaussian correctly handles warmup period and NaN propagation"""
        close = test_data['close']
        period = 14
        poles = 4

        result = ta_indicators.gaussian(close, period=period, poles=poles)


        assert len(result) == len(close)


        assert np.all(np.isfinite(result)), "Expected all finite values for clean input data"


        close_with_nan = np.copy(close)
        nan_start = 100
        nan_end = 105
        close_with_nan[nan_start:nan_end] = np.nan

        result_with_nan = ta_indicators.gaussian(close_with_nan, period=period, poles=poles)


        assert np.all(np.isfinite(result_with_nan[:nan_start])), f"Expected finite values before NaN region"



        warmup_end = nan_end + period
        assert np.all(np.isnan(result_with_nan[nan_start:warmup_end])), f"Expected NaN values during and after NaN input"




    def test_gaussian_streaming(self, test_data):
        """Test Gaussian streaming vs batch calculation - mirrors check_gaussian_streaming"""
        close = test_data['close'][:100]
        period = 14
        poles = 4


        batch_result = ta_indicators.gaussian(close, period, poles)


        stream = ta_indicators.GaussianStream(period, poles)
        stream_results = []

        for val in close:
            result = stream.update(val)
            stream_results.append(result)

        stream_results = np.array(stream_results)


        assert_close(
            stream_results[period:],
            batch_result[period:],
            rtol=1e-9,
            msg="Gaussian streaming vs batch mismatch"
        )

    def test_gaussian_batch(self, test_data):
        """Test Gaussian batch computation."""
        close = test_data['close']


        period_range = (10, 20, 5)
        poles_range = (2, 4, 1)

        result = ta_indicators.gaussian_batch(close, period_range, poles_range)

        assert isinstance(result, dict)
        assert 'values' in result
        assert 'periods' in result
        assert 'poles' in result

        expected_periods = [10, 10, 10, 15, 15, 15, 20, 20, 20]
        expected_poles = [2, 3, 4, 2, 3, 4, 2, 3, 4]

        assert list(result['periods']) == expected_periods
        assert list(result['poles']) == expected_poles
        assert result['values'].shape == (9, len(close))


        row_idx = 0
        for period in [10, 15, 20]:
            for poles in [2, 3, 4]:
                individual_result = ta_indicators.gaussian(close, period, poles)
                np.testing.assert_allclose(
                    result['values'][row_idx],
                    individual_result,
                    rtol=1e-9,
                    err_msg=f"Batch row {row_idx} (period={period}, poles={poles}) mismatch"
                )
                row_idx += 1

    def test_gaussian_different_poles(self, test_data):
        """Test Gaussian with different poles values."""
        close = test_data['close']
        period = 14


        for poles in [1, 2, 3, 4]:
            result = ta_indicators.gaussian(close, period, poles)
            assert len(result) == len(close)



            assert len(result) == len(close)

            assert not np.isnan(result[period:]).any()

    def test_gaussian_kernel_parameter(self, test_data):
        """Test that kernel parameter works correctly"""
        close = test_data['close']
        period = 14
        poles = 4


        result_auto = ta_indicators.gaussian(close, period, poles)
        result_scalar = ta_indicators.gaussian(close, period, poles, kernel='scalar')


        valid_idx = ~np.isnan(result_auto)
        np.testing.assert_allclose(result_auto[valid_idx], result_scalar[valid_idx], rtol=1e-10)


        with pytest.raises(ValueError):
            ta_indicators.gaussian(close, period, poles, kernel='invalid')


        batch_result = ta_indicators.gaussian_batch(
            close,
            period_range=(10, 20, 5),
            poles_range=(2, 4, 1),
            kernel='scalar'
        )
        assert 'values' in batch_result
        assert 'periods' in batch_result
        assert 'poles' in batch_result